### PR TITLE
IGNITE-16351 C++ thin, ODBC: SSL certificates, keys and CA are not mandatory

### PR DESCRIPTION
**Pull request by isapego**
- Improved error reporting during SSL failures.
- Fixed issue when client compiled and run with different OpenSSL versions.
- Re-factored to remove SSL-related code duplication.
- Added missing tests to some test suites.
- Certificate, private key and CA are not mandatory parameters for client now. Also, implemented extracting and usage of system CA for Windows by default.

### The Contribution Checklist
- [x] There is a single JIRA ticket related to the pull request. 
- [x] The web-link to the pull request is attached to the JIRA ticket.
- [x] The JIRA ticket has the _Patch Available_ state.
- [x] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [x] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [x] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [x] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

**Review by ivandasch — APPROVED** [review]
Looks good! Few minor comments added, please take a look

File: modules/platforms/cpp/common/src/common/utils.cpp [hunk 1]
```diff
@@ -229,5 +231,46 @@ namespace ignite
...
+
+            if (!advice.empty())
+                messageBuilder << ". " << advice;
+
+            return messageBuilder.str();
+        }
+
+        std::string GetLastSystemError(const std::string& description, const std::string& advice)
+        {
+            return common::FormatErrorMessage(description, GetLastSystemError(), advice);
```
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/common/src/common/utils.cpp:266]: Fixed.

File: modules/platforms/cpp/network/os/linux/src/network/linux_async_worker_thread.h [hunk 2]
```diff
@@ -135,13 +135,6 @@ namespace ignite
              */
             bool ShouldInitiateNewConnection() const;
 
-            /**
```
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/network/os/linux/src/network/linux_async_worker_thread.h:138]: Removed.

File: modules/platforms/cpp/network/src/network/ssl/ssl_gateway.cpp [hunk 3]
```diff
@@ -218,9 +220,7 @@ namespace ignite
                 functions.fpSSL_write = LoadSslMethod("SSL_write");
                 functions.fpSSL_read = LoadSslMethod("SSL_read");
                 functions.fpSSL_pending = LoadSslMethod("SSL_pending");
-                functions.fpSSL_get_state = TryLoadSslMethod("SSL_get_state");
-                if (!functions.fpSSL_get_state)
-                    functions.fpSSL_get_state = LoadSslMethod("SSL_state");
+                functions.fpSSL_get_version = LoadSslMethod("SSL_get_version");
```
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/network/src/network/ssl/ssl_gateway.cpp:223]: Fixed

File: modules/platforms/cpp/network/os/linux/src/network/linux_async_client_pool.cpp [hunk 4]
```diff
@@ -112,7 +110,16 @@ namespace ignite
 
             bool closed = client.Get()->Close();
             if (closed)
+            {
+                IgniteError err0(client.Get()->GetCloseError());
```
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/network/os/linux/src/network/linux_async_client_pool.cpp:114]: This is not a hot path and to me copying here is safer than operating on reference.

[the remark this answers is not in the document]
Comment (ivandasch) [reply modules/platforms/cpp/network/os/linux/src/network/linux_async_client_pool.cpp:114]: I meant that we can rewrite code
```
if (!err)
                {
                    const IgniteError& baseErr = client.Get()->GetCloseError();
                    std::auto_ptr<IgniteError> err0;
                    if (baseErr.GetCode() == IgniteError::IGNITE_SUCCESS)
                    {
                        err0 = std::auto_ptr<IgniteError>(new IgniteError(
                                IgniteError::IGNITE_ERR_NETWORK_FAILURE, "Connection closed by server")
                        );
                    }
                    else
                        err0 = std::auto_ptr<IgniteError>(new IgniteError(baseErr));

                    HandleConnectionClosed(id, err0.get());
                }
                else
                    HandleConnectionClosed(id, err);
```

**Pull request closed without merging**